File: java/simplifypath.java
public class Solution {
    
    
    
    public String simplifyPath(String path) {
        // Start typing your Java solution below
        // DO NOT write main() function
        if(path.length()==1) return path;
        path = path.replaceAll("/$","");
        path = path.replaceAll("/+","/");
        if(path.length()==1) return path;
        Stack<String> s = new Stack<String>();
        while(path.length()!=0){
            int index = path.indexOf('/');
            if(index==-1) index = path.length();
            if(index==0) {
                path = path.substring(1);
                continue;
            }
            String tmp = path.substring(0,index);
            if(index!=path.length()){
                path = path.substring(index+1);
            }else{
                path = "";
            }
            
            if(tmp.equals(".")){
                
            }else if(tmp.equals("..")){
                if(!s.empty()){
                    s.pop();
                }
            }else{
                s.push(tmp);
            }
        }
        String ans = "";
        while(!s.empty()){
            ans = "/" + s.pop() + ans;
        }
        if(ans.length()==0) ans = "/";
        return ans;
        
    }
}

File: cpp/SimplifyPath.cpp
class Solution {
public:
    string simplifyPath(string path) {
        //path += '/';
        vector<string> s;
        for(int i=0;i<path.length();) {
            if(path[i] == '/') {
                while(path[i]=='/') {
                    i++;
                }
                continue;
            }
            string folder = "";
            while( i<path.length() && path[i]!='/') {
                folder.append(1, path[i]);
                i++;
            }
            if(folder==".") {
                
            } else if(folder=="..") {
                if(!s.empty()) {
                    s.pop_back();
                }
            } else {
                s.push_back(folder);
            }
        }
        if(s.empty()) {
            return "/";
        }
        string res = "";
        for(int i=0;i<s.size();i++) {
            res += ("/" + s[i]);
        }
        return res;
    }
};
